pkg/services/ngalert/remote/alertmanager.go

```diff
@@ -389,6 +389,48 @@ func (am *Alertmanager) sendConfiguration(ctx context.Context, cfg *remoteClient
 	return nil
 }
 
+// RemoteState represents the state (silences, nflog) in use by a remote Alertmanager.
+type RemoteState struct {
+	Silences []byte
+	Nflog    []byte
+}
+
+// GetRemoteState gets the remote Alertmanager's internal state.
+func (am *Alertmanager) GetRemoteState(ctx context.Context) (RemoteState, error) {
+	var rs RemoteState
+
+	s, err := am.mimirClient.GetFullState(ctx)
+	if err != nil {
+		return rs, fmt.Errorf("failed to pull remote state: %w", err)
+	}
+
+	// Decode and unmarshal the base64-encoded state we got from Mimir.
+	decoded, err := base64.StdEncoding.DecodeString(s.State)
+	if err != nil {
+		return rs, fmt.Errorf("failed to base64-decode remote state: %w", err)
+	}
+	protoState := &alertingClusterPB.FullState{}
+	if err := protoState.Unmarshal(decoded); err != nil {
+		return rs, fmt.Errorf("failed to unmarshal remote state: %w", err)
+	}
+
+	// Mimir state has two parts:
+	// - "sil:<tenantID>": silences
+	// - "nfl:<tenantID>": notification log entries
+	for _, p := range protoState.Parts {
+		switch p.Key {
+		case "sil:" + am.tenantID:
+			rs.Silences = p.Data
+		case "nfl:" + am.tenantID:
+			rs.Nflog = p.Data
+		default:
+			return rs, fmt.Errorf("unknown part key %q", p.Key)
+		}
+	}
```

Comment:
I would rather do it on the server and return the "parsed" state instead but considering short lifetime of this code, I think I am good with this approach

Reply:
I'm trying to be consistent and make Mimir return the full state in the same format it uses to [return the saved Grafana Alertmanager state](https://github.com/grafana/mimir/blob/63db22387abe78f0e0e9d2662deeed41517966e0/pkg/alertmanager/api_grafana.go#L196-L207). Grafana also sends its state to Mimir as a base64-encoded string.